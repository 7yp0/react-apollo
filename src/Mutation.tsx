import * as React from 'react';
import ApolloClient, { PureQueryOptions, ApolloError } from 'apollo-client';
import { DataProxy } from 'apollo-cache';
import { DocumentNode, GraphQLError } from 'graphql';
import { ApolloConsumer as Consumer } from './Context';

const invariant = require('invariant');
const shallowEqual = require('fbjs/lib/shallowEqual');

import { OperationVariables, RefetchQueriesProviderFn } from './types';
import { parser, DocumentType } from './parser';

export interface MutationResult<TData = Record<string, any>> {
  data?: TData;
  error?: ApolloError;
  loading: boolean;
  called: boolean;
}

export interface ExecutionResult<T = Record<string, any>> {
  data?: T;
  extensions?: Record<string, any>;
  errors?: GraphQLError[];
}

// Improved MutationUpdaterFn type, need to port them back to Apollo Client
export declare type MutationUpdaterFn<
  T = {
    [key: string]: any;
  }
> = (proxy: DataProxy, mutationResult: FetchResult<T>) => void;

export declare type FetchResult<C = Record<string, any>, E = Record<string, any>> = ExecutionResult<
  C
> & {
  extensions?: E;
  context?: C;
};

export declare type MutationOptions<TData = any, TVariables = OperationVariables> = {
  variables?: TVariables;
  optimisticResponse?: Object;
  refetchQueries?: string[] | PureQueryOptions[] | RefetchQueriesProviderFn;
  update?: MutationUpdaterFn<TData>;
};

export declare type MutationFn<TData = any, TVariables = OperationVariables> = (
  options?: MutationOptions<TData, TVariables>,
) => Promise<void | FetchResult<TData>>;

export interface MutationProps<TData = any, TVariables = OperationVariables> {
  mutation: DocumentNode;
  ignoreResults?: boolean;
  optimisticResponse?: Object;
  variables?: TVariables;
  refetchQueries?: string[] | PureQueryOptions[] | RefetchQueriesProviderFn;
  update?: MutationUpdaterFn<TData>;
  children: (
    mutateFn: MutationFn<TData, TVariables>,
    result: MutationResult<TData>,
  ) => React.ReactNode;
  onCompleted?: (data: TData) => void;
  onError?: (error: ApolloError) => void;
  context?: Record<string, any>;
}

export interface MutationState<TData = any> {
  called: boolean;
  error?: ApolloError;
  data?: TData;
  loading: boolean;
  client: ApolloClient;
}

const initialState = {
  loading: false,
  called: false,
  error: undefined,
  data: undefined,
  client: undefined,
};

class Mutation<TData = any, TVariables = OperationVariables> extends React.Component<
  MutationProps<TData, TVariables>,
  MutationState<TData>
> {
  private mostRecentMutationId: number;

  private hasMounted: boolean;

  constructor(props: MutationProps<TData, TVariables>) {
    super(props);

    this.verifyDocumentIsMutation(props.mutation);

    this.mostRecentMutationId = 0;
    this.state = { called: false, client: props.client, loading: false };
  }

  componentDidMount() {
    this.hasMounted = true;
  }

  componentWillUnmount() {
    this.hasMounted = false;
  }

  static getDerivedStateFromProps = (nextProps, prevState) => {
    if (nextProps.client !== prevState.client) {
      return {
        ...initialState,
        client: nextProps.client,
      };
    }
    return null;
  };

  componentDidUpdate(prevProps: InnerMutationProps<TData, TVariables>) {
    if (this.props.mutation !== prevProps.mutation) {
      this.verifyDocumentIsMutation(this.props.mutation);
    }
  }

  render() {
    const { children } = this.props;
    const { loading, data, error, called } = this.state;

    const result = {
      called,
      loading,
      data,
      error,
    };

    return children(this.runMutation, result);
  }

  private runMutation = (options: MutationOptions<TVariables> = {}) => {
    this.onStartMutation();

    const mutationId = this.generateNewMutationId();

    return this.mutate(options)
      .then(response => {
        this.onCompletedMutation(response, mutationId);
        return response;
      })
      .catch(e => {
        this.onMutationError(e, mutationId);
        if (!this.props.onError) throw e;
      });
  };

  private mutate = (options: MutationOptions<TVariables>) => {
    const { mutation, variables, optimisticResponse, update, context = {} } = this.props;
    let refetchQueries = options.refetchQueries || this.props.refetchQueries;
    return this.state.client.mutate({
      mutation,
      variables,
      optimisticResponse,
      refetchQueries,
      update,
      context,
      ...options,
    });
  };

  private onStartMutation = () => {
    if (!this.state.loading && !this.props.ignoreResults) {
      this.setState({
        loading: true,
        error: undefined,
        data: undefined,
        called: true,
      });
    }
  };

  private onCompletedMutation = (response: ExecutionResult<TData>, mutationId: number) => {
    if (this.hasMounted === false) {
      return;
    }
    const { onCompleted, ignoreResults } = this.props;

    const data = response.data as TData;

    const callOncomplete = () => (onCompleted ? onCompleted(data) : null);

    if (this.isMostRecentMutation(mutationId) && !ignoreResults) {
      this.setState({ loading: false, data }, callOncomplete);
    } else {
      callOncomplete();
    }
  };

  private onMutationError = (error: ApolloError, mutationId: number) => {
    if (this.hasMounted === false) {
      return;
    }
    const { onError } = this.props;
    const callOnError = () => (onError ? onError(error) : null);

    if (this.isMostRecentMutation(mutationId)) {
      this.setState({ loading: false, error }, callOnError);
    } else {
      callOnError();
    }
  };

  private generateNewMutationId = (): number => {
    this.mostRecentMutationId = this.mostRecentMutationId + 1;
    return this.mostRecentMutationId;
  };

  private isMostRecentMutation = (mutationId: number) => {
    return this.mostRecentMutationId === mutationId;
  };

  private verifyDocumentIsMutation = (mutation: DocumentNode) => {
    const operation = parser(mutation);
    invariant(
      operation.type === DocumentType.Mutation,
      `The <Mutation /> component requires a graphql mutation, but got a ${
        operation.type === DocumentType.Query ? 'query' : 'subscription'
      }.`,
    );
  };
}

export default class ApolloMutation<
  TData = any,
  TVariables = OperationVariables
> extends React.Component<MutationProps<TData, TVariables>> {
  render() {
    return <Consumer>{client => <Mutation client={client} {...this.props} />}</Consumer>;
  }
}
