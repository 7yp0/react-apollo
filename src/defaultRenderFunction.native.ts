export default function(_: React.ReactNode): string {
  throw new Error("Please use getMarkupFromTree with a custom renderFunction in React Native");
}
