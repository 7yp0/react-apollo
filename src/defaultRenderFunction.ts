export {
  renderToStaticMarkup as default
} from "react-dom/server";
