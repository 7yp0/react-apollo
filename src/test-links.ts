import {
  Operation,
  GraphQLRequest,
  ApolloLink,
  FetchResult,
  Observable,
  // Observer,
} from 'apollo-link';

import { print } from 'graphql/language/printer';
import { makeExecutableSchema, addMockFunctionsToSchema, IMocks } from 'graphql-tools';
import { addTypenameToDocument } from 'apollo-utilities';
import { graphql, DocumentNode, buildClientSchema, printSchema } from 'graphql';
const isEqual = require('lodash.isequal');

export interface MockedResponse {
  request: GraphQLRequest;
  result?: FetchResult;
  error?: Error;
  delay?: number;
  newData?: () => FetchResult;
}

export interface MockedSubscriptionResult {
  result?: FetchResult;
  error?: Error;
  delay?: number;
}

export interface MockedSubscription {
  request: GraphQLRequest;
}

export class MockLink extends ApolloLink {
  public addTypename: Boolean = true;
  private mockedResponsesByKey: { [key: string]: MockedResponse[] } = {};

  constructor(mockedResponses: ReadonlyArray<MockedResponse>, addTypename: Boolean = true) {
    super();
    this.addTypename = addTypename;
    if (mockedResponses)
      mockedResponses.forEach((mockedResponse, index) => {
        if (!mockedResponse.request.query)
          console.warn(`Mock at index ${index} is missing a 'query'`);
        this.addMockedResponse(mockedResponse);
      });
  }

  public addMockedResponse(mockedResponse: MockedResponse) {
    const key = requestToKey(mockedResponse.request, this.addTypename);
    let mockedResponses = this.mockedResponsesByKey[key];
    if (!mockedResponses) {
      mockedResponses = [];
      this.mockedResponsesByKey[key] = mockedResponses;
    }
    mockedResponses.push(mockedResponse);
  }

  public request(operation: Operation) {
    const key = requestToKey(operation, this.addTypename);
    let responseIndex;
    const response = (this.mockedResponsesByKey[key] || []).find((res, index) => {
      const requestVariables = operation.variables || {};
      const mockedResponseVariables = res.request.variables || {};
      if (!isEqual(requestVariables, mockedResponseVariables)) {
        return false;
      }
      responseIndex = index;
      return true;
    });

    if (!response || typeof responseIndex === 'undefined') {
      throw new Error(
        `No more mocked responses for the query: ${print(
          operation.query,
        )}, variables: ${JSON.stringify(operation.variables)}`,
      );
    }

    this.mockedResponsesByKey[key].splice(responseIndex, 1);

    const { result, error, delay, newData } = response;

    if (newData) {
      response.result = newData();
      this.mockedResponsesByKey[key].push(response);
    }

    if (!result && !error) {
      throw new Error(`Mocked response should contain either result or error: ${key}`);
    }

    return new Observable<FetchResult>(observer => {
      let timer = setTimeout(
        () => {
          if (error) {
            observer.error(error);
          } else {
            if (result) observer.next(result);
            observer.complete();
          }
        },
        delay ? delay : 0,
      );

      return () => {
        clearTimeout(timer);
      };
    });
  }
}

export class MockSubscriptionLink extends ApolloLink {
  // private observer: Observer<any>;
  public unsubscribers: any[] = [];
  public setups: any[] = [];

  private observer: any;

  constructor() {
    super();
  }

  public request(_req: any) {
    return new Observable<FetchResult>(observer => {
      this.setups.forEach(x => x());
      this.observer = observer;
      return () => {
        this.unsubscribers.forEach(x => x());
      };
    });
  }

  public simulateResult(result: MockedSubscriptionResult) {
    setTimeout(() => {
      const { observer } = this;
      if (!observer) throw new Error('subscription torn down');
      if (result.result && observer.next) observer.next(result.result);
      if (result.error && observer.error) observer.error(result.error);
    }, result.delay || 0);
  }

  public onSetup(listener: any): void {
    this.setups = this.setups.concat([listener]);
  }

  public onUnsubscribe(listener: any): void {
    this.unsubscribers = this.unsubscribers.concat([listener]);
  }
}

function requestToKey(request: GraphQLRequest, addTypename: Boolean): string {
  const queryString =
    request.query && print(addTypename ? addTypenameToDocument(request.query) : request.query);

  const requestKey = { query: queryString };

  return JSON.stringify(requestKey);
}

// Pass in multiple mocked responses, so that you can test flows that end up
// making multiple queries to the server
// NOTE: The last arg can optionally be an `addTypename` arg
export function mockSingleLink(...mockedResponses: Array<any>): ApolloLink {
  // to pull off the potential typename. If this isn't a boolean, we'll just set it true later
  let maybeTypename = mockedResponses[mockedResponses.length - 1];
  let mocks = mockedResponses.slice(0, mockedResponses.length - 1);

  if (typeof maybeTypename !== 'boolean') {
    mocks = mockedResponses;
    maybeTypename = true;
  }

  return new MockLink(mocks, maybeTypename);
}

export function mockObservableLink(): MockSubscriptionLink {
  return new MockSubscriptionLink();
}

/**
 * AutoMockLink
 */

export interface AutoMockLinkArgs {
  // schema: string | DocumentNode;
  schema: any;
  addTypename?: Boolean;
  mocks?: IMocks;
}

export class AutoMockLink extends ApolloLink {
  public addTypename?: Boolean;
  private executableSchema: any;

  // schema: string || DocumentAST
  constructor({ schema, addTypename = true, mocks }: AutoMockLinkArgs) {
    super();
    this.addTypename = addTypename;

    const schemaString =
      schema && schema.__schema ? printSchema(buildClientSchema(schema)) : schema;

    const executableSchema = makeExecutableSchema({ typeDefs: schemaString });
    addMockFunctionsToSchema({ schema: executableSchema, mocks });
    this.executableSchema = executableSchema;
  }

  public request(operation: Operation) {
    const result = graphql({
      schema: this.executableSchema,
      source: print(operation.query),
      variableValues: operation.variables,
    });

    return new Observable<FetchResult>(observer => {
      result.then(({ errors, data }) => {
        if (errors) observer.error(errors);
        else {
          if (data) observer.next({ data });
          observer.complete();
        }
      });
    });
  }
}
