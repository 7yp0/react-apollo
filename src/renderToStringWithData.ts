import { ReactElement } from 'react';
import defaultRenderFunction from './defaultRenderFunction';
import { default as getDataFromTree } from './getDataFromTree';

export function renderToStringWithData(component: ReactElement<any>): Promise<string> {
  return getDataFromTree(component).then(() => defaultRenderFunction(component));
}
