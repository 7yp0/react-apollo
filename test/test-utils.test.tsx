import * as React from 'react';
import * as renderer from 'react-test-renderer';
import { InMemoryCache } from 'apollo-cache-inmemory';
import ApolloClient from 'apollo-client';
import gql from 'graphql-tag';

import { graphql, ChildProps, Mutation, Query } from '../src';
import { MockedProvider, mockSingleLink, AutoMockedProvider } from '../src/test-utils';
import { MockedResponse } from '../src/test-links';
import { DocumentNode, isNullableType } from 'graphql';

import * as fs from 'fs';
import path from 'path';

// get the json file and parse it to a js object
const jsonSchema = JSON.parse(
  fs.readFileSync(path.resolve(__dirname, 'fixtures/schema.json'), {
    encoding: 'utf-8',
  }),
);

// import { execute, buildSchema, introspectionQuery } from 'graphql';

const variables = {
  username: 'mock_username',
};

const userWithoutTypeName = {
  id: 'user_id',
};

const user = {
  __typename: 'User',
  ...userWithoutTypeName,
};

const query: DocumentNode = gql`
  query GetUser($username: String!) {
    user(username: $username) {
      id
    }
  }
`;
const queryWithTypename: DocumentNode = gql`
  query GetUser($username: String!) {
    user(username: $username) {
      id
      __typename
    }
  }
`;

interface Data {
  user: {
    id: string;
  };
}

interface Variables {
  username: string;
}

const withUser = graphql<Variables, Data, Variables>(query, {
  options: props => ({
    variables: props,
  }),
});

const mocks: ReadonlyArray<MockedResponse> = [
  {
    request: {
      query,
      variables,
    },
    result: { data: { user } },
  },
];

it('mocks the data', done => {
  class Container extends React.Component<ChildProps<Variables, Data, Variables>> {
    componentWillReceiveProps(nextProps: ChildProps<Variables, Data, Variables>) {
      expect(nextProps.data!.user).toMatchSnapshot();
      done();
    }

    render() {
      return null;
    }
  }

  const ContainerWithData = withUser(Container);

  renderer.create(
    <MockedProvider mocks={mocks}>
      <ContainerWithData {...variables} />
    </MockedProvider>,
  );
});

it('allows for querying with the typename', done => {
  class Container extends React.Component<ChildProps<Variables, Data, Variables>> {
    componentWillReceiveProps(nextProps: ChildProps<Variables, Data, Variables>) {
      expect(nextProps.data!.user).toMatchSnapshot();
      done();
    }

    render() {
      return null;
    }
  }

  const withUserAndTypename = graphql<Variables, Data, Variables>(queryWithTypename, {
    options: props => ({
      variables: props,
    }),
  });

  const ContainerWithData = withUserAndTypename(Container);

  const mocksWithTypename = [
    {
      request: {
        query: queryWithTypename,
        variables,
      },
      result: { data: { user } },
    },
  ];

  renderer.create(
    <MockedProvider mocks={mocksWithTypename}>
      <ContainerWithData {...variables} />
    </MockedProvider>,
  );
});

it('allows for using a custom cache', done => {
  const cache = new InMemoryCache();
  cache.writeQuery({
    query,
    variables,
    data: { user },
  });

  const Container: React.SFC<ChildProps<Variables, Data, Variables>> = props => {
    expect(props.data).toMatchObject({ user });
    done();

    return null;
  };
  const ContainerWithData = withUser(Container);
  renderer.create(
    <MockedProvider mocks={[]} cache={cache}>
      <ContainerWithData {...variables} />
    </MockedProvider>,
  );
});

it('errors if the variables in the mock and component do not match', done => {
  class Container extends React.Component<ChildProps<Variables, Data, Variables>> {
    componentWillReceiveProps(nextProps: ChildProps<Variables, Data, Variables>) {
      try {
        expect(nextProps.data!.user).toBeUndefined();
        expect(nextProps.data!.error).toMatchSnapshot();
        done();
      } catch (e) {
        done.fail(e);
      }
    }

    render() {
      return null;
    }
  }

  const ContainerWithData = withUser(Container);

  const variables2 = {
    username: 'other_user',
  };

  renderer.create(
    <MockedProvider mocks={mocks}>
      <ContainerWithData {...variables2} />
    </MockedProvider>,
  );
});

it('errors if the variables do not deep equal', done => {
  class Container extends React.Component<ChildProps<Variables, Data, Variables>> {
    componentWillReceiveProps(nextProps: ChildProps<Variables, Data, Variables>) {
      try {
        expect(nextProps.data!.user).toBeUndefined();
        expect(nextProps.data!.error).toMatchSnapshot();
        done();
      } catch (e) {
        done.fail(e);
      }
    }

    render() {
      return null;
    }
  }

  const ContainerWithData = withUser(Container);

  const mocks2 = [
    {
      request: {
        query,
        variables: {
          age: 13,
          username: 'some_user',
        },
      },
      result: { data: { user } },
    },
  ];

  const variables2 = {
    username: 'some_user',
    age: 42,
  };

  renderer.create(
    <MockedProvider mocks={mocks2}>
      <ContainerWithData {...variables2} />
    </MockedProvider>,
  );
});

it('does not error if the variables match but have different order', done => {
  class Container extends React.Component<ChildProps<Variables, Data, Variables>> {
    componentWillReceiveProps(nextProps: ChildProps<Variables, Data, Variables>) {
      expect(nextProps.data!.user).toMatchSnapshot();
      done();
    }

    render() {
      return null;
    }
  }

  const ContainerWithData = withUser(Container);

  const mocks2 = [
    {
      request: {
        query,
        variables: {
          age: 13,
          username: 'some_user',
        },
      },
      result: { data: { user } },
    },
  ];

  const variables2 = {
    username: 'some_user',
    age: 13,
  };

  renderer.create(
    <MockedProvider mocks={mocks2}>
      <ContainerWithData {...variables2} />
    </MockedProvider>,
  );
});

it('mocks a network error', done => {
  class Container extends React.Component<ChildProps<Variables, Data, Variables>> {
    componentWillReceiveProps(nextProps: ChildProps<Variables, Data, Variables>) {
      try {
        expect(nextProps.data!.error).toEqual(new Error('Network error: something went wrong'));
        done();
      } catch (e) {
        done.fail(e);
      }
    }

    render() {
      return null;
    }
  }

  const ContainerWithData = withUser(Container);

  const mocksError = [
    {
      request: {
        query,
        variables,
      },
      error: new Error('something went wrong'),
    },
  ];

  renderer.create(
    <MockedProvider mocks={mocksError}>
      <ContainerWithData {...variables} />
    </MockedProvider>,
  );
});

it('errors if the query in the mock and component do not match', done => {
  class Container extends React.Component<ChildProps<Variables, Data, Variables>> {
    componentWillReceiveProps(nextProps: ChildProps<Variables, Data, Variables>) {
      try {
        expect(nextProps.data!.user).toBeUndefined();
        expect(nextProps.data!.error).toMatchSnapshot();
        done();
      } catch (e) {
        done.fail(e);
      }
    }

    render() {
      return null;
    }
  }

  const ContainerWithData = withUser(Container);

  const mocksDifferentQuery = [
    {
      request: {
        query: gql`
          query OtherQuery {
            otherQuery {
              id
            }
          }
        `,
        variables,
      },
      result: { data: { user } },
    },
  ];

  renderer.create(
    <MockedProvider mocks={mocksDifferentQuery}>
      <ContainerWithData {...variables} />
    </MockedProvider>,
  );
});

it('doesnt crash on unmount if there is no query manager', () => {
  class Container extends React.Component {
    render() {
      return null;
    }
  }

  renderer
    .create(
      <MockedProvider>
        <Container />
      </MockedProvider>,
    )
    .unmount();
});

describe('AutoMockedProvider', () => {
  const schema = `
    type Query {
      user(username: String!): User!
    }

    type Mutation {
      echo(message: String!): String!
    }
    
    type User {
      id: ID!
    }    
  `;

  const mocks = {
    User: () => ({
      id: () => 'harambe',
    }),
  };

  const mutation = gql`
    mutation echo($message: String!) {
      echo(message: $message)
    }
  `;

  it('allows for querying with autoMocks', done => {
    class Container extends React.Component<ChildProps<Variables, Data, Variables>> {
      componentWillReceiveProps(nextProps: ChildProps<Variables, Data, Variables>) {
        expect(nextProps.data!.user!.id).toBeDefined();
        done();
      }

      render() {
        return null;
      }
    }

    const withUserAndTypename = graphql<Variables, Data, Variables>(query, {
      options: props => ({
        variables: props,
      }),
    });

    const ContainerWithData = withUserAndTypename(Container);

    renderer.create(
      <AutoMockedProvider schema={schema}>
        <ContainerWithData {...variables} />
      </AutoMockedProvider>,
    );
  });

  it('allows for manual mocks', done => {
    class Container extends React.Component<ChildProps<Variables, Data, Variables>> {
      componentWillReceiveProps(nextProps: ChildProps<Variables, Data, Variables>) {
        expect(nextProps.data!.user!.id).toEqual('harambe');
        done();
      }

      render() {
        return null;
      }
    }

    const withUserAndTypename = graphql<Variables, Data, Variables>(query, {
      options: props => ({
        variables: props,
      }),
    });

    const ContainerWithData = withUserAndTypename(Container);

    renderer.create(
      <AutoMockedProvider schema={schema} mocks={mocks}>
        <ContainerWithData {...variables} />
      </AutoMockedProvider>,
    );
  });

  it('allows for mutations', done => {
    renderer.create(
      <AutoMockedProvider schema={schema} mocks={mocks}>
        <Mutation
          mutation={mutation}
          onCompleted={({ echo }) => {
            expect(echo).toEqual('Hello World');
            done();
          }}
        >
          {(mutate, { data, loading, error }) => {
            if (!data && !loading && !error) {
              mutate({ variables: { message: 'ohh hai' } });
            }

            return null;
          }}
        </Mutation>
      </AutoMockedProvider>,
    );
  });

  it('works with json schemas', done => {
    class Container extends React.Component<ChildProps<Variables, Data, Variables>> {
      componentWillReceiveProps(nextProps: ChildProps<Variables, Data, Variables>) {
        expect(nextProps.data!.user!.id).toEqual('harambe');
        done();
      }

      render() {
        return null;
      }
    }

    const withUserAndTypename = graphql<Variables, Data, Variables>(query, {
      options: props => ({
        variables: props,
      }),
    });

    const ContainerWithData = withUserAndTypename(Container);

    // THIS WILL WORK WHEN AC 2.x bugs are fixed
    // const TestComponent = () => (
    //   <Query query={query}>
    //     {({ data }) => {
    //       console.log({ data });
    //       if (data && data.user) {
    //         expect(data!.user!.id).toEqual('harambe');
    //         done();
    //       }
    //       return null;
    //     }}
    //   </Query>
    // );

    renderer.create(
      <AutoMockedProvider schema={jsonSchema} mocks={mocks}>
        <ContainerWithData {...variables} />
        {/* <TestComponent /> */}
      </AutoMockedProvider>,
    );
  });
});
