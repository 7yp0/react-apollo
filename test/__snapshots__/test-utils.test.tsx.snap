// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MockedProvider allows for querying with the typename 1`] = `
Object {
  "__typename": "User",
  "id": "user_id",
  Symbol(id): "User:user_id",
}
`;

exports[`MockedProvider does not error if the variables match but have different order 1`] = `
Object {
  "__typename": "User",
  "id": "user_id",
  Symbol(id): "User:user_id",
}
`;

exports[`MockedProvider errors if the query in the mock and component do not match 1`] = `
[Error: Network error: No more mocked responses for the query: query GetUser($username: String!) {
  user(username: $username) {
    id
    __typename
  }
}
, variables: {"username":"mock_username"}]
`;

exports[`MockedProvider errors if the variables do not deep equal 1`] = `
[Error: Network error: No more mocked responses for the query: query GetUser($username: String!) {
  user(username: $username) {
    id
    __typename
  }
}
, variables: {"username":"some_user","age":42}]
`;

exports[`MockedProvider errors if the variables in the mock and component do not match 1`] = `
[Error: Network error: No more mocked responses for the query: query GetUser($username: String!) {
  user(username: $username) {
    id
    __typename
  }
}
, variables: {"username":"other_user"}]
`;

exports[`MockedProvider mocks the data 1`] = `
Object {
  "__typename": "User",
  "id": "user_id",
  Symbol(id): "User:user_id",
}
`;
